File: 1 kyu/BECOME IMMORTAL/solution.kt
import java.math.BigInteger
/**
 * set true to enable debug
 */
internal var debug = false

internal fun elderAge(n: Long, m: Long, k: Long, newp: Long): Long {
    fun calculate(n: Long, m: Long, k: Long, delta: Long): BigInteger {
        fun intervalSum(a: Long, b: Long): BigInteger {
            if (b < a) return BigInteger.ZERO
            val bigA = BigInteger.valueOf(a)
            val bigB = BigInteger.valueOf(b)
            return ((bigA+bigB)*(bigB-bigA+BigInteger.ONE)).shiftRight(1)
        }
        if (m > n) return calculate(m, n, k, delta)
        val nH = java.lang.Long.highestOneBit(n)
        val mH = if (m < nH) m else nH
        var result = intervalSum(Math.max(0, delta - k), delta + nH - k - 1) * BigInteger.valueOf(mH)
        if (n > nH) result = (result + calculate(n - nH, mH, k, delta + nH))
        if (m > mH) result = (result + calculate(nH, m - mH, k, delta + mH))
        if (n > nH && m > mH) result = (result + calculate(n - nH, m - mH, k, delta + (mH xor nH)))
        return result
    }
    return calculate(n, m, k, 0).mod(BigInteger.valueOf(newp)).toLong()
}
__________________________________
internal var debug = false

internal fun elderAge(m: Long, n: Long, l: Long, t: Long): Long {
  val base: Double = 2.toDouble()
  val small: Long = Math.min(m, n)
  val big: Long = Math.max(m, n)
  val power: Long = Math.pow(base, Math.floor(Math.log(big.toDouble()) / Math.log(base)).toDouble()).toLong()
  val rows: Long = Math.min(power, small)
  val first: Long = Math.max(0, -l)
  val terms: Long = Math.max(0, power - l - 1)
  var x = terms - first + 1
  var y = first + terms

  if (x % 2 > 0) y = Math.floor((y / 2).toDouble()).toLong()
  else if (y % 2 > 0) x = Math.floor((x / 2).toDouble()).toLong()

  val series: Long = if (y <= 0) 0 else ((y % t) * (x % t)) % t

  var sum = ((series % t) * (rows % t)) % t
  if (big > power) sum += elderAge(big - power, rows, l - power, t)
  if (small > rows) sum += elderAge(power, small - rows, l - rows, t)
  if (small > rows && big > power) sum += elderAge(big - power, small - rows, l, t)

  return sum % t
}
____________________________________________
internal var debug = false

internal fun mmul(x: Long, y: Long, m: Long) : Long = (x % m) * (y % m) % m

internal fun elderAge(m: Long, n: Long, l: Long, t: Long): Long {
    if (m < n) return elderAge(n, m, l, t)
    val k1 = 2 * m.takeHighestOneBit()
    if (m <= 0 || n <= 0 || l >= k1) return 0
    val k = if (n > 1) k1 / 2 else m
    val r = Math.min(n, k)
    var s = 0L
    if (k > l) {
        if (l >= 0) {
            var x = k - l
            var y = x - 1
            if (x % 2 == 0L) x /= 2 else y /= 2
            s = mmul(mmul(x, y, t), r, t)
        } else {
            var x = k
            var y = k - 1
            if (x % 2 == 0L) x /= 2 else y /= 2
            s = mmul(mmul(x, y, t) - mmul(l, k, t), r, t)
        }
    }
    s += elderAge(m - k, r, l - k, t)
    s += elderAge(k, n - r, l - k, t)
    s += elderAge(m - k, n - r, l, t)
    return s % t
}
____________________________________________
import kotlin.math.*

val debug = true

fun elderAge(row: Long, col: Long, loss: Long, newp: Long): Long {
    val biggerDim = max(row, col)
    val smallerDim = min(row, col)
    return elderAge(0, biggerDim, smallerDim, loss, newp)
}


private fun elderAge(
    offset: Long,
    biggerDim: Long,
    smallerDim: Long,
    loss: Long,
    mod: Long
): Long {
    val smallerPow2 = highestPowerOf2(smallerDim)
    val biggerPow2 = highestPowerOf2(biggerDim)

    val mainPart = modulatedSeqSum(
        start = max(0, offset - loss),
        end = max(0, offset + biggerPow2 - 1 - loss),
        length = max(1, biggerPow2 + min(0, offset - loss)),
        times = smallerPow2,
        mod = mod
    )

    val smallerSidePart = modulatedSeqSum(
        start = max(0, offset + smallerPow2 - loss),
        end = max(0, offset + (smallerPow2 * 2) - 1 - loss),
        length = max(1, smallerPow2 + min(0, offset + smallerPow2 - loss)),
        times = max(0, smallerDim - smallerPow2),
        mod = mod
    )

    var biggerSidePart = 0L
    if (biggerDim > biggerPow2 && biggerPow2 < 2 * smallerPow2) {
        biggerSidePart = modulatedSeqSum(
            start = max(0, offset + smallerPow2 - loss),
            end = max(0, offset + (smallerPow2 * 2) - 1 - loss),
            length = max(1, smallerPow2 + min(0, offset + smallerPow2 - loss)),
            times = max(0, min(biggerDim - smallerPow2, smallerPow2)),
            mod = mod
        )
    }

    var bottomReminderPart = 0L
    if (smallerDim > smallerPow2 && biggerPow2 > smallerPow2 * 2) {
        bottomReminderPart = modulatedSeqSum(
            start = max(0, offset + 2 * smallerPow2 - loss),
            end = max(0, offset + biggerPow2 - 1 - loss),
            length = max(0, biggerPow2 - 2 * smallerPow2 + min(0, offset + 2 * smallerPow2 - loss)),
            times = max(0, smallerDim - smallerPow2),
            mod = mod
        )
    }

    var mirroredPart = 0L
    if (smallerDim > smallerPow2 && biggerDim > smallerPow2) {
        mirroredPart = elderAge(
            offset = offset,
            biggerDim = max(0, min(biggerDim - smallerPow2, smallerPow2)),
            smallerDim = max(0, smallerDim - smallerPow2), // negative numbers here should never happen
            loss = loss,
            mod = mod
        )
    }


    var remainderPart = 0L
    if (biggerDim > biggerPow2 && biggerPow2 >= 2 * smallerPow2) {
        remainderPart = elderAge(
            offset = offset + biggerPow2,
            biggerDim = max(biggerDim - biggerPow2, smallerDim),
            smallerDim = min(biggerDim - biggerPow2, smallerDim),
            loss = loss,
            mod = mod
        )
    }
    return (mainPart + smallerSidePart + biggerSidePart + mirroredPart + remainderPart + bottomReminderPart) % mod
}

// if is needed because division modulo (by 0.5) is not always defined
fun modulatedSeqSum(start: Long, end: Long, length: Long, times: Long, mod: Long): Long =
    if ((start + end) % 2 == 0L) {
        (((0.5 * start % mod + 0.5 * end % mod) % mod) * (((length % mod) * (times % mod)) % mod) % mod).toLong()
    } else {
        ((((0.5 * length) % mod * (times % mod)) % mod * (start % mod + end % mod) % mod) % mod).toLong()
    }

fun highestPowerOf2(n: Long): Long {
    val p = (ln(n.toDouble()) / ln(2.0)).toLong()
    return 2.0.pow(p.toDouble()).toLong()
}

File: 1 kyu/BECOME IMMORTAL/solution.dart
59568be9cc15b57637000054

import 'dart:math';

/** set true to enable debug */
bool debug = false;

BigInt max(BigInt a, BigInt b) { return a > b ? a : b; }
BigInt clog2(BigInt a) { BigInt b = BigInt.one; while (b < a) b *= BigInt.two; return b; }
BigInt rsum(BigInt a, BigInt b) { return ((a + b) * (b - a + BigInt.one)) ~/ BigInt.two; }

int elderAge(int m, int n, int k, int p) {
  return helper(BigInt.from(m), BigInt.from(n), BigInt.from(k), BigInt.from(p)).toInt();
}

BigInt helper(BigInt m, BigInt n, BigInt k, BigInt p) {
  if (m == BigInt.zero || n == BigInt.zero) return BigInt.zero;
  if (m > n) { BigInt o = m; m = n; n = o; }
  BigInt x = clog2(m);
  BigInt y = clog2(n);
  if (k > y) return BigInt.zero;
  if (x == y) return ((m + n - y) * rsum(BigInt.one, y - k - BigInt.one) + helper(y - n, x - m, k, p)) % p;
  x = y ~/ BigInt.two;
  BigInt phi = m * rsum(BigInt.one, y - k - BigInt.one) - (y - n) * rsum(max(BigInt.zero, x - k), y - k - BigInt.one);
  if (k <= x) phi += (x - k) * (x - m) * (y - n) + helper(x - m, y - n, BigInt.zero, p);
  else phi += helper(x - m, y - n, k - x, p);
  return phi % p;
}
______________________________________
bool debug = false;

int mul(int x, int y, int m) {
  return (x % m) * (y % m) % m;
}

int elderAge(int m, int n, int l, int t) {
  if (m < n) return elderAge(n, m, l, t);
  var k = 1 << m.bitLength;
  if (m <= 0 || n <= 0 || l >= k) return 0;
  k = n > 1 ? k >> 1 : m;
  var r = n < k ? n : k, s = 0;
  if (k > l) {
    if (l >= 0) {
      var x = k - l;
      var y = x - 1;
      if (x.isEven) x ~/= 2; else y ~/= 2;
      s = mul(mul(x, y, t), r, t);
    }
    else {
      var x = k, y = k - 1;
      if (x.isEven) x ~/= 2; else y ~/= 2;
      s = mul(mul(x, y, t) - mul(l, k, t), r, t);
    }
  }
  s += elderAge(m - k, r, l - k, t);
  s += elderAge(k, n - r, l - k, t);
  s += elderAge(m - k, n - r, l, t);
  return s % t;
}
______________________________________________
import 'dart:math' as math;

/** set true to enable debug */
bool debug = false;

BigInt bigIntPow(BigInt x) {
  var p = BigInt.one;
  while (p < x) {
    p <<= 1;
  }
  return p;
}

BigInt bigIntRangeSum(BigInt l, BigInt r) =>
    (l + r) * (r - l + BigInt.one) ~/ BigInt.two;

BigInt bigIntMax(BigInt a, BigInt b) => a > b ? a : b;

BigInt bigIntElderAge(BigInt m, BigInt n, BigInt k, BigInt newp) {
  if (m == BigInt.zero || n == BigInt.zero) {
    return BigInt.zero;
  }
  if (m > n) {
    var t = m;
    m = n;
    n = t;
  }
  var lm = bigIntPow(m);
  var ln = bigIntPow(n);
  if (k > ln) {
    return BigInt.zero;
  }

  if (lm == ln) {
    return ((bigIntRangeSum(BigInt.one, ln - k - BigInt.one) * (m + n - ln) +
            bigIntElderAge(ln - n, lm - m, k, newp))) %
        newp;
  }
  if (lm < ln) {
    lm = ln ~/ BigInt.from(2);
    var tmp = bigIntRangeSum(BigInt.one, ln - k - BigInt.one) * m -
        (ln - n) *
            bigIntRangeSum(bigIntMax(BigInt.zero, lm - k), ln - k - BigInt.one);

    if (k <= lm) {
      tmp += (lm - k) * (lm - m) * (ln - n) +
          bigIntElderAge(lm - m, ln - n, BigInt.zero, newp);
    } else {
      tmp += bigIntElderAge(lm - m, ln - n, k - lm, newp);
    }
    return tmp % newp;
  }
  return BigInt.zero;
}

int elderAge(int m, int n, int k, int newp) => bigIntElderAge(
        BigInt.from(m), BigInt.from(n), BigInt.from(k), BigInt.from(newp))
    .toInt();
______________________________________________
import 'dart:math';

bool debug = false;

int elderAge(int n, int m, int k, int newp) {
  var findMax = max(n, m);
  var findMin = min(n, m);

  var sq = pow(2.0, (log(findMax.toDouble()) / log(2)).floor()).toInt();

  var sMin = min(sq, findMin);
  var kInv = max(0, -k);
  var maxMod = max(0, sq - k - 1);

  var dCase = maxMod - kInv + 1;
  var sCase = maxMod + kInv;
  if (dCase % 2 > 0) {
    sCase = (sCase.toDouble() / 2).floor().toInt();
  } else if (sCase % 2 > 0) {
    dCase = (dCase.toDouble() / 2).floor().toInt();
  }

  var diff;
  if (sCase <= 0) {
    diff = 0;
  } else {
    diff = ((sCase % newp) * (dCase % newp)) % newp;
  }
  var acc = ((diff % newp) * (sMin % newp)) % newp;
  if (findMax > sq) {
    acc += elderAge(findMax - sq, sMin, k - sq, newp);
  }
  if (findMin > sMin) {
    acc += elderAge(sq, findMin - sMin, k - sMin, newp);
  }
  if ((findMin > sMin) & (findMax > sq)) {
    acc += elderAge(findMax - sq, findMin - sMin, k, newp);
  }
  return acc % newp;
}
